wcfsetup/install/files/lib/http/middleware/CheckUserBan.class.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+namespace wcf\http\middleware;
+
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\ServerRequestInterface;
+use Psr\Http\Server\MiddlewareInterface;
+use Psr\Http\Server\RequestHandlerInterface;
+use wcf\data\user\User;
+use wcf\http\Helper;
+use wcf\system\exception\AJAXException;
+use wcf\system\exception\NamedUserException;
+use wcf\system\WCF;
+
+/**
+ * Checks whether the user is banned and deletes their sessions.
+ *
+ * @author  Tim Duesterhus
+ * @copyright   2001-2022 WoltLab GmbH
+ * @license GNU Lesser General Public License <http://opensource.org/licenses/lgpl-license.php>
+ * @package WoltLabSuite\Core\Http\Middleware
+ * @since   6.0
+ */
+final class CheckUserBan implements MiddlewareInterface
+{
+    /**
+     * @inheritDoc
+     */
+    public function process(ServerRequestInterface $request, RequestHandlerInterface $handler): ResponseInterface
+    {
+        $user = WCF::getUser();
+
+        if ($this->isBanned($user)) {
+            if (Helper::isAjaxRequest($request)) {
+                throw new AJAXException(
+                    WCF::getLanguage()->getDynamicVariable('wcf.user.error.isBanned'),
+                    AJAXException::INSUFFICIENT_PERMISSIONS
+                );
+            } else {
+                // Delete sessions only for non-AJAX requests to ensure
+                // that the user was able to see the message properly
+                WCF::getSession()->deleteUserSessionsExcept($user);
+
+                throw new NamedUserException(WCF::getLanguage()->getDynamicVariable('wcf.user.error.isBanned'));
+            }
+        }
+
+        return $handler->handle($request);
+    }
+
+    private function isBanned(User $user): bool
```

Comment:
Wouldn't it be better to move this code to `User.class.php`?

Reply:
Yes, it would. Looking through the code, this should be done in a follow-up PR, because various locations already check for `->banned` without taking owner access into account.

Reply:
see #4938